File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.cpp
//[[1,100,3],[7,8,9]], k = 2
//max(solve(indexOfPile, remainingCoins ))



class Solution {
public:
    
    vector<vector<int>> dp; // dp[index][remaining coins]
    int solve(vector<vector<int>> &piles, int index, int remainCoins )
    {
        if(index >= piles.size()  || remainCoins<=0)
            return 0;
        //without taking from indexth pile
        if(dp[index][remainCoins]!=-1)
            return dp[index][remainCoins];
        int not_take = solve(piles, index+1, remainCoins);
        int ans = not_take;
        //taking from indexth  pile
        int x = piles[index].size();
        int range = min(remainCoins,x);
        int take =INT_MIN;
        int sum =0;
        for(int i=1;i<=range;i++)
        {
            sum+= piles[index][i-1];
            ans = max(ans,sum + solve(piles, index+1, remainCoins-i));
        }
                    
        return dp[index][remainCoins] = ans;
    }
    int maxValueOfCoins(vector<vector<int>>& piles, int k) {
        int n = piles.size();
        dp.resize(n+1, vector<int>(k+1, -1));
        return solve(piles,0 , k);
       
    }
};

File: 2218-maximum-value-of-k-coins-from-piles/2218-maximum-value-of-k-coins-from-piles.java
class Solution {
    public int maxValueOfCoins(List<List<Integer>> piles, int k) {
        int n = piles.size();
        int dp[][] = new int[n+1][k+1];
        for(int i=0;i<dp.length;i++)
            Arrays.fill(dp[i], -1);
        int ans = solve(piles, 0, k, dp); 
        return ans;
    }
    
    //solve functions gives the max sum that can be formed till indexth pile and
    //remainCoins
    private int solve(List<List<Integer>> piles, int index , int remainCoins , 
                     int [][] dp)
    {
        if(index>=piles.size() || remainCoins<=0)
            return 0;
        
        if(dp[index][remainCoins] != -1 )
            return dp[index][remainCoins];
        
        int ans = 0;
        int not_take = solve(piles, index+1 , remainCoins , dp);
        ans = not_take;
        int sum = 0;
        int range = Math.min(remainCoins, piles.get(index).size());
        for(int i=1;i<=range;i++)
        {
            sum+=piles.get(index).get(i-1);
            ans = Math.max(ans, sum + solve(piles, index + 1, remainCoins -i, dp));
        }
        
        return dp[index][remainCoins] = ans;
        
    }
}